tests/test_dynamics_client.py

```diff
@@ -205,108 +205,185 @@ def test_client_headers(dynamics_client):
 
 
 def test_client_init_from_environment():
-    os.environ["DYNAMICS_API_URL"] = "apiurl"
-    os.environ["DYNAMICS_TOKEN_URL"] = "tokenurl"
-    os.environ["DYNAMICS_CLIENT_ID"] = "clientid"
-    os.environ["DYNAMICS_CLIENT_SECRET"] = "secret"
-    os.environ["DYNAMICS_SCOPE"] = "scope"
+    complete_set = [
+        ("DYNAMICS_API_URL", "apiurl"),
+        ("DYNAMICS_TOKEN_URL", "tokenurl"),
+        ("DYNAMICS_CLIENT_ID", "clientid"),
+        ("DYNAMICS_CLIENT_SECRET", "secret"),
+        ("DYNAMICS_SCOPE", "scope"),
+    ]
 
+    # minimum required including scope
+    for key, val in complete_set:
+        os.environ[key] = val
     try:
         with mock.patch("dynamics.client.DynamicsClient.get_token"):
             client = DynamicsClient.from_environment()
+            assert client._api_url == "apiurl/"
+            assert client._session._client.client_id == "clientid"
     except KeyError as error:
         pytest.fail(f"Wrong environment variables: {error}")
-        return
 
-    assert client._api_url == "apiurl/"
-    assert client._session._client.client_id == "clientid"
+    # test required env vars individually
```

Comment:
These commented sections should be individual tests, so an individual test only has one reason to fail.